### PR TITLE
Types cdata wrapping

### PR DESCRIPTION
**Pull request by exvito**
@opalmer 

Here is the first take at the PR. General notes:
- I followed the strategy generally outlined in the mailing list: https://groups.google.com/forum/#!topic/pywincffi/B9JNULyvhgI
- I did lots of small commits, in what I believe is a mostly sane order (thus, to better understand the PR, it may be useful to look at a few commits at a time).
- General result:
  - The end user APIs no longer expose CFFI <cdata> objects: whether arguments or results.
  - Full separation/clarification of what is text and what is binary at the API wrapper level.
- Details:
  - C structs are now wrapped in simple, light but customizable Python objects.
  - HANDLEs are also wrapped; they're even lighter than structs, and need special handling, but I thing the API utilization results in a much nicer experice.
  - Type validation is simplified and the input_check function reflects that.
  - Given strict text/binary separation string_to_cdata no longer needed.
  - Some tests had to be adjusted to the revamped types / APIs.
  - Some no longer made sense and I threw them away.
- Doc strings will need to be reviewed.
- Tests are passing here.
- There will certainly things that need fixing, of course.

Your input is welcome.
Cheers.

### CODE REVIEW

**Comment by exvito** [comment]
Note:
- Some tests are failing on Python 3.x -- I'll go ahead and address that next. (no Py 3 environment here!)
- I'm not sure how to "fix" some of the "PYLINT=1" complaints -- I'll try and improve on that as well.
- The "READTHEDOCS=1" failure looks like a mystery -- I'll leave that for last.

I forgot to mention:
- I used one cffi 1.6.0 API (having had to upgrade my venv) in kernel32/file.py on the last return statement of ReadFile: its ffi.unpack as documented here http://cffi.readthedocs.io/en/latest/ref.html#ffi-string-ffi-unpack

**Comment by exvito** [comment]
Further notes:
- I guess I finally fixed the Python 3 tests and pylint errors/warnings.
- What I need your expertise for: addressing the READTHEDOCS=1 failures which involve tech I'm not familiar with at all (travis + sphinx). I find it odd, that 2.x and 3.x versions seem to be failing differently.

Waiting on your input.

File: pywincffi/core/typesbase.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+Provides the base types on top of which user visible types will be built.
+"""
+
+import cffi
+
+
+# pylint: disable=too-few-public-methods
+class CFFICDataWrapper(object):
+
```
Comment (opalmer) [inline pywincffi/core/typesbase.py]: Shouldn't have newlines before the doc string starts.

File: pywincffi/core/typesbase.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    """
+    Base class for exposing Python types and interfaces to pywincffi users.
+    * Wraps a CFFI cdata object in self._cdata.
+    * Delegates attribute getting/setting to self._cdata, supporting structs.
+    * Delegates item getting/setting to self._cdata, supporting arrays.
+    Attribute access is not delegated to the wrapped object if the class
```
Comment (opalmer) [inline pywincffi/core/typesbase.py:18]: nit: newline before the paragraph starts

File: pywincffi/core/typesbase.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+"""
+
+import cffi
+
+
+# pylint: disable=too-few-public-methods
+class CFFICDataWrapper(object):
+
+    """
+    Base class for exposing Python types and interfaces to pywincffi users.
```
Comment (opalmer) [inline pywincffi/core/typesbase.py]: nit: newline before the bullet list starts, probably would be nice to have : at the end of this line too.

File: pywincffi/core/typesbase.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    * Delegates attribute getting/setting to self._cdata, supporting structs.
+    * Delegates item getting/setting to self._cdata, supporting arrays.
+    Attribute access is not delegated to the wrapped object if the class
+    itself contains such an attribute and that attribute is a descriptor; this
+    is in place to support @property in sub-classes.
+    """
+
+    def __init__(self, cdecl, ffi=None):
+        if ffi is None:
+            ffi = cffi.FFI()
```
Comment (opalmer) [inline pywincffi/core/typesbase.py]: You'll want to use `dist.load()` here to retrieve the ffi object:

``` python
if ffi is None:
    ffi, _ = dist.load()
```

dist.py caches an instance of `cffi.FFI` which may be loaded from the precompiled version of pywincffi.

File: pywincffi/core/typesbase.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        self._cdata = ffi.new(cdecl)
+
+    def __getattr__(self, name):
+        return getattr(self._cdata, name)
+
+    def __setattr__(self, name, value):
+        # avoid self-recursion setting own attribute: use parent's __setattr__
+        if name == '_cdata':
```
Comment (opalmer) [inline pywincffi/core/typesbase.py]: ' -> " (for consistency)

File: pywincffi/core/typesbase.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+class CFFICDataWrapper(object):
+
+    """
+    Base class for exposing Python types and interfaces to pywincffi users.
+    * Wraps a CFFI cdata object in self._cdata.
+    * Delegates attribute getting/setting to self._cdata, supporting structs.
+    * Delegates item getting/setting to self._cdata, supporting arrays.
+    Attribute access is not delegated to the wrapped object if the class
+    itself contains such an attribute and that attribute is a descriptor; this
+    is in place to support @property in sub-classes.
```
Comment (opalmer) [inline pywincffi/core/typesbase.py:20]: Could you add some parameter and keyword documentation for cdecl and ffi?  Does not have to be anything complex, just enough to describe the inputs.

File: pywincffi/kernel32/events.py [hunk 7]
```diff
@@ -34,7 +34,8 @@ def CreateEvent(
         A pointer to a ``SECURITY_ATTRIBUTES`` structure.  If not provided
         then by default the handle cannot be inherited by a subprocess.
 
-    :keyword str lpName:
+    :keyword unicode/str lpName:
```
Comment (opalmer) [inline pywincffi/kernel32/events.py]: I was going to comment on this and suggest:

```
:type lpName: unicode or str
:keyword lpName:
```

or 

```
:keyword str lpName:
```

After playing around with it some though, I don't think it matters (so this is more of an FYI I guess).  If you put just `str` in there sphinx will create a link to the standard string type.  It seems it allows `unicode/str` and the `:type ...` form I have above but neither of those create links back for the standard types.

File: pywincffi/kernel32/events.py [hunk 8]
```diff
@@ -88,24 +86,25 @@ def OpenEvent(dwDesiredAccess, bInheritHandle, lpName):
         The access desired for the event object.
 
     :param bool bInheritHandle:
-    :param str lpName:
+    :param unicode/str lpName:
+        Type is unicode on Python 2, str on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/events.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/kernel32/events.py [hunk 9]
```diff
@@ -34,7 +34,8 @@ def CreateEvent(
         A pointer to a ``SECURITY_ATTRIBUTES`` structure.  If not provided
         then by default the handle cannot be inherited by a subprocess.
 
-    :keyword str lpName:
+    :keyword unicode/str lpName:
+        Type is unicode on Python 2, str on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/events.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/kernel32/file.py [hunk 10]
```diff
@@ -27,7 +29,8 @@ def CreateFile(  # pylint: disable=too-many-arguments
         https://msdn.microsoft.com/en-us/library/aa363858
         https://msdn.microsoft.com/en-us/library/gg258116
 
-    :param str lpFileName:
+    :param unicode/str lpFileName:
+        Type is unicode on Python 2, str on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/file.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/kernel32/file.py [hunk 11]
```diff
@@ -131,66 +125,53 @@ def WriteFile(hFile, lpBuffer, nNumberOfBytesToWrite=None, lpOverlapped=None,
     :param handle hFile:
         The handle to write to
 
-    :type lpBuffer: bytes, string or unicode.
+    :type lpBuffer: str/bytes
     :param lpBuffer:
-        The data to be written to the file or device. We should be able
-        to convert this value to unicode.
+        Type is str on Python 2, bytes on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/file.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/kernel32/file.py [hunk 12]
```diff
@@ -131,66 +125,53 @@ def WriteFile(hFile, lpBuffer, nNumberOfBytesToWrite=None, lpOverlapped=None,
...
-        >>> hFile = None # normally, this would be a handle
-        >>> lpOverlapped = ffi.new(
-        ...     "OVERLAPPED[1]", [{
-        ...         "hEvent": hFile
-        ...     }]
-        ... )
+        >>> from pywincffi.kernel32 import WriteFile, CreateEvent
+        >>> from pywincffi.wintypes import OVERLAPPED
+        >>> hEvent = CreateEvent(...)
+        >>> lpOverlapped = OVERLAPPED()
```
Comment (opalmer) [inline pywincffi/kernel32/file.py:145]: Thanks for updating the example both here and below 👍

File: pywincffi/kernel32/file.py [hunk 13]
```diff
@@ -274,10 +254,12 @@ def MoveFileEx(lpExistingFileName, lpNewFileName, dwFlags=None):
 
         https://msdn.microsoft.com/en-us/library/aa365240
 
-    :param str lpExistingFileName:
+    :param unicode/str lpExistingFileName:
+        Type is unicode on Python 2, str on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/file.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/kernel32/file.py [hunk 14]
```diff
@@ -274,10 +254,12 @@ def MoveFileEx(lpExistingFileName, lpNewFileName, dwFlags=None):
...
         Name of the file or directory to perform the operation on.
 
-    :param str lpNewFileName:
+    :param unicode/str lpNewFileName:
+        Type is unicode on Python 2, str on Python 3.
```
Comment (opalmer) [inline pywincffi/kernel32/file.py]: I'd rephrase so it's clear the doc string is talking about the objects:

```
Type is ``unicode`` on Python 2, ``str`` on Python 3.
```

You could also do:

```
Type is :class:`unicode` on Python 2, :class:`str` on Python 3.
```

I'd suggest the first version though because it render identically in Python 2 and 3 (docs could be built in either version).

File: pywincffi/wintypes/__init__.py [hunk 15]
```diff
@@ -0,0 +1,113 @@
+"""
+Windows Types Package
+====================
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]: The footer (===) needs to be as long as the section header:

```
"""
Windows Types Package
=====================

Provides user accessible types corresponding to the respective Windows types
used across the exposed APIs.
"""
```

File: pywincffi/wintypes/__init__.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+"""
+
+
+from six import integer_types
+
+from pywincffi.core import typesbase
+from pywincffi.core import dist
+
+
+_FFI, _ = dist.load()
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]: So this is why the doc build steps fail.

When you call `dist.load()` it will either import `_pywincffi` or try to build it if it does not exist.  Travis builds the documentation on Linux just like [readthedocs.org](https://readthedocs.org/projects/pywincffi/builds/) across multiple Python versions so if there's a problem like this it's easy to discover while doing development versus after the fact.  Of course Linux does not have access the headers and libraries that Windows does so this in turn causes the build to fail.

I suggest using `dist.load()` inside of each of the functions in this module like we do elsewhere in pywincffi.  This prevents pywincffi from being built and allows static tools like sphinx and pylint to run on any platform.

File: pywincffi/wintypes/__init__.py [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+
+# pylint: disable=protected-access
+def wintype_to_cdata(wintype):
+    """
+    :param wintype:
+        A type derived from :class:`pywincffi.core.typesbase.CFFICDataWrapper`
+    :return:
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py:34]: Newline before `:return:`

File: pywincffi/wintypes/__init__.py [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+from pywincffi.core import dist
+
+
+_FFI, _ = dist.load()
+
+
+# pylint: disable=protected-access
+def wintype_to_cdata(wintype):
+    """
+    :param wintype:
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py:31]: Could we get a brief description of the function's purpose and maybe a small example?

File: pywincffi/wintypes/__init__.py [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    elif isinstance(wintype, HANDLE):
+        return wintype._cdata[0]
+    else:
+        return wintype._cdata
+
+
+# pylint: disable=too-few-public-methods
+class HANDLE(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/aa383751
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]:
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/aa383751
```

File: pywincffi/wintypes/__init__.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+
+    def __repr__(self):
+        return "<HANDLE 0x%x at 0x%x>" % (
+            int(_FFI.cast("intptr_t", self._cdata[0])),
+            id(self)
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, HANDLE):
+            raise TypeError('%r must be a HANDLE' % other)
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]: ' -> " (for consistency)

File: pywincffi/wintypes/__init__.py [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+        return self._cdata[0] == other._cdata[0]
+
+
+# pylint: disable=invalid-name
+class SECURITY_ATTRIBUTES(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/aa379560
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]:
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/aa379560
```

File: pywincffi/wintypes/__init__.py [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    def nLength(self, _):
+        # Can't raise AttributeError.
+        # CFFICDataWrapper would fall back to setting self._cdata.nLength .
+        raise TypeError("SECURITY_ATTRIBUTES.nLength is read-only")
+
+
+# pylint: disable=too-few-public-methods
+class OVERLAPPED(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/ms684342
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]:
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/ms684342
```

File: pywincffi/wintypes/__init__.py [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+    # pylint: disable=missing-docstring
+    @property
+    def hEvent(self):
+        return HANDLE(self._cdata.hEvent)
+
+    # pylint: disable=missing-docstring
+    @hEvent.setter
+    def hEvent(self, handle):
+        if not isinstance(handle, HANDLE):
+            raise TypeError('%r must be a HANDLE' % handle)
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]: ' -> " (for consistency)

File: pywincffi/wintypes/__init__.py [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+        self._cdata.hEvent = handle._cdata[0]
+
+
+# pylint: disable=too-few-public-methods
+class FILETIME(typesbase.CFFICDataWrapper):
+    """
+    .. seealso: https://msdn.microsoft.com/en-us/library/ms724284
```
Comment (opalmer) [inline pywincffi/wintypes/__init__.py]:
```
.. seealso:

    https://msdn.microsoft.com/en-us/library/ms724284
```

**Comment by opalmer** [comment]
Did a review and overall it looks great, thanks again for putting the time into this.  Based on the changes in the tests around input/output type handling this definitely looks like it will be a major improvement once it lands.

Most of my comments where around consistency, doc strings, etc so nothing major to really fix.  I plan to do a deeper dive and play around with the new types some myself before a final review but I don't think that will end up changing anything.  

So there's really only two fixes here that break things and need to be fixed:
- You need to call dist.load() inside of the functions in typebases.py
- Some doc string formatting which could result in odd rendering issues.
- Update the cffi requirement in setup.py to `cffi>=1.6.0`.  As you stated in your comment above we're using `ffi.unpack` which came out in 1.6.0.  Since we rely on that API we have to make sure downstream consumers of pywincffi are using the proper cffi version.

File: pywincffi/core/typesbase.py (hunk 6 above)
Comment (exvito) [inline pywincffi/core/typesbase.py:20]: Making ffi a non-optional arg. Benefit: Ensure loose coupling between this class and the remaining code.

Notes:
- wintypes was already providing the dist.load() ffi instance.
- test_typesbase with minor review, explicitly passing "dummy" ffi instance.

File: pywincffi/wintypes/__init__.py (hunk 16 above)
Comment (exvito) [inline pywincffi/wintypes/__init__.py]: Thanks for clarifying. Changing this.

**Comment by exvito** [comment]
I believe I addressed all of your suggestions with the prior commit.
Other than any further review you may share, I think the following actions are still pending:
- Removing the pywincffi.util module - string_to_cdata no longer used.
- Reviewing docstrings in the wrappers.
- Updating the change log.
- Not sure - Maybe the docs need review?

File: tests/test_core/test_typesbase.py [hunk 25]
```diff
@@ -4,12 +4,15 @@
...
+_ffi = cffi.FFI()
+
+
 class TestCFFICDataWrapper(TestCase):
     """
     Tests for :class:`pywincffi.core.typesbase.CFFICDataWrapper`
     """
     def test_instantiate(self):
-        o = typesbase.CFFICDataWrapper("char *")
+        o = typesbase.CFFICDataWrapper("char *", _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:15]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 26]
```diff
@@ -19,7 +22,7 @@ def test_simple_struct_set_and_get(self):
                 float y;
                 float z;
             } *
-        """)
+        """, _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:25]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 27]
```diff
@@ -28,7 +31,7 @@ def test_simple_struct_set_and_get(self):
         self.assertAlmostEqual(o.z, 789.0, places=2)
 
     def test_char_array_set_and_get(self):
-        o = typesbase.CFFICDataWrapper("char [256]")
+        o = typesbase.CFFICDataWrapper("char [256]", _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:34]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 28]
```diff
@@ -43,16 +46,15 @@ def test_char_array_set_and_get(self):
             self.assertEqual(o[i], content[start:finish])
 
     def test_char_array_set_and_ffi_string(self):
-        ffi = cffi.FFI()
-        o = typesbase.CFFICDataWrapper("char [256]")
+        o = typesbase.CFFICDataWrapper("char [256]", _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:49]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 29]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
                     } socket;
                 } u;
             } *
-        """)
+        """, _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:93]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 30]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
         _uuid = b"01234567-abcd-abcd-abcd-0123456789ab"
         o.uuid = _uuid
         o.fd = 0xfd
         o.type = 2
         o.u.file.inode = 12345
         _file_name = b"the-filename.test"
         o.u.file.name = _file_name
-        self.assertEqual(ffi.unpack(o.uuid, len(_uuid)), _uuid)
+        self.assertEqual(_ffi.unpack(o.uuid, len(_uuid)), _uuid)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:101]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 31]
```diff
@@ -61,7 +63,7 @@ def test_simple_struct_array_set_and_get(self):
                 int minute;
                 int second;
             } [20]
-        """)
+        """, _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:66]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 32]
```diff
@@ -43,16 +46,15 @@ def test_char_array_set_and_get(self):
...
+        o = typesbase.CFFICDataWrapper("char [256]", _ffi)
         content = b"0123456789"
         content_len = len(content)
         for i in range(content_len):
             # Have Python 3 return a len()==1 bytes object
             o[i] = content[i:i+1]
         # must use wrapped cdata to apply ffi.string in this case
         # won't be needed for struct members as other tests verify
-        self.assertEqual(ffi.string(o._cdata), content)
+        self.assertEqual(_ffi.string(o._cdata), content)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:57]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 33]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
         _file_name = b"the-filename.test"
         o.u.file.name = _file_name
-        self.assertEqual(ffi.unpack(o.uuid, len(_uuid)), _uuid)
+        self.assertEqual(_ffi.unpack(o.uuid, len(_uuid)), _uuid)
         self.assertEqual(o.fd, 0xfd)
         self.assertEqual(o.type, 2)
         self.assertEqual(o.u.file.inode, 12345)
         self.assertEqual(
-            ffi.unpack(o.u.file.name, len(_file_name)),
+            _ffi.unpack(o.u.file.name, len(_file_name)),
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:106]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 34]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
             _file_name
         )
 
     def test_externally_provided_ffi(self):
-        ffi = cffi.FFI()
-        ffi.cdef("""
+        _ffi.cdef("""
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:111]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 35]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
             typedef struct _person {
                 wchar_t first_name[32];
                 wchar_t last_name[32];
             } person_t;
-        """)
-        o = typesbase.CFFICDataWrapper("person_t *", ffi)
+        """, _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:116]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 36]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
+        o = typesbase.CFFICDataWrapper("person_t *", _ffi)
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:117]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 37]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
         o.first_name = u"First Name"
         o.last_name = u"Last Name"
-        self.assertEqual(ffi.string(o.first_name), u"First Name")
-        self.assertEqual(ffi.string(o.last_name), u"Last Name")
+        self.assertEqual(_ffi.string(o.first_name), u"First Name")
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:120]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

File: tests/test_core/test_typesbase.py [hunk 38]
```diff
@@ -89,36 +90,35 @@ def test_complex_struct_set_and_get(self):
...
+        self.assertEqual(_ffi.string(o.last_name), u"Last Name")
```
Comment (opalmer) [inline tests/test_core/test_typesbase.py:121]: Should use dist.load() to retrieve an instance of ffi like we do in normal code.

**Comment by opalmer** [comment]
Changes look good, thanks.  Had a few comments on the tests though but it's basically all the same comment: we should use dist.load() in the tests like we do in normal code.

I also agree with your list of TODOs, especially the doc string review (which is something I intend to take another pass at myself too).

**Comment by exvito** [comment]
My idea is to keep pywincffi/core/types.base (CFFICDataWrapper) + associated tests (test_typesbase.py you commented on above) completely uncoupled from the rest of the code. That's why those tests use independent FFI instances. In other words: the class and associated tests have zero references/dependencies on the remaining code, they should only depend on CFFI.

On the other hand, the actual types in use (pywincffi/wintypes/**init**.py) do have their tests dependent on the project code and, in particular, in the FFI object obtained from dist.load() in tests/test_wintypes. Those are "implementing" C types declared in the header files that dist.load() handles and need to be tested against that.

If you don't agree with this approach, I can change that, of course. But I think keeping the CFFICDataWrapper + tests with no dependencies is a cleaner design with no drawbacks that I can identify now, while future-proofing against future changes (which will be simpler with less dependencies).

Let me know of your thoughts.

**Comment by opalmer** [comment]
Alright that seems fine to me, thanks for explaining. I took another look at the docs and unless you have further changes I think I'd be ok with merging this.

Thanks again! 👍

**Comment by exvito** [comment]
Completed:
- Reviewed doc strings and updated types.
- Removed the no longer used pywincffi/util.py containing the string_to_cdata function.
- Updated changelog.rst.

For some reason not clear to me, the tests here are failing complaining that this branch has conflicts. Why it won't run the tests I don't understand (even though there may be merge conflicts, of course).

Other than that, I agree with you, and if tests come out ok, I think this is ready to be merged.

**Comment by exvito** [comment]
Investigated a bit and found out that the merge conflict came down to a simple thing in docs/source/changelog.rst. Merged in your master, fixed changelog.rst manually and away it goes.

Let's see if the tests come out ok for travis and appveyor.

**Comment by opalmer** [comment]
Looks good, not seeing anything additional here that needs to be fixed so I'll go ahead and merge this.  This also fixes #86.  Thanks again, looking forward to using these changes in future PRs!
